src/libmongoc/tests/test-mongoc-crud.c

```diff
@@ -51,8 +51,135 @@ test_all_spec_tests (TestSuite *suite)
       suite, resolved, &test_crud_cb, TestSuite_CheckLive);
 }
 
+static void
+prose_test_1 (void *ctx)
+{
+   mongoc_client_t *client;
+   mongoc_collection_t *coll;
+   bool ret;
+   bson_t reply;
+   bson_error_t error;
+
+   client = test_framework_new_default_client ();
+   coll = get_test_collection (client, "coll");
+
+   ret = mongoc_client_command_simple (
+      client,
+      "admin",
+      tmp_bson ("{'configureFailPoint': 'failCommand', 'mode': {'times': 1}, "
+                " 'data': {'failCommands': ['insert'], 'writeConcernError': {"
+                "   'code': 100, 'codeName': 'UnsatisfiableWriteConcern', "
+                "   'errmsg': 'Not enough data-bearing nodes', "
+                "   'errInfo': {'writeConcern': {'w': 2, 'wtimeout': 0, "
+                "               'provenance': 'clientSupplied'}}}}}"),
+      NULL,
+      NULL,
+      &error);
+   ASSERT_OR_PRINT (ret, error);
+
+   ret = mongoc_collection_insert_one (
+      coll, tmp_bson ("{'x':1}"), NULL /* opts */, &reply, &error);
+   ASSERT (!ret);
+
+   /* libmongoc does not model WriteConcernError, so we only assert that the
+    * "errInfo" field set in configureFailPoint matches that in the result */
+   ASSERT_MATCH (&reply,
+                 "{'writeConcernErrors': [{'errInfo': {'writeConcern': {"
+                 "'w': 2, 'wtimeout': 0, 'provenance': 'clientSupplied'}}}]}");
+
+   bson_destroy (&reply);
+   mongoc_collection_destroy (coll);
+   mongoc_client_destroy (client);
+}
+
+typedef struct {
+   bool has_reply;
+   bson_t reply;
+} prose_test_2_apm_ctx_t;
+
+static void
+prose_test_2_command_succeeded (const mongoc_apm_command_succeeded_t *event)
+{
+   if (!strcmp (mongoc_apm_command_succeeded_get_command_name (event),
+                "insert")) {
+      prose_test_2_apm_ctx_t *ctx =
+         mongoc_apm_command_succeeded_get_context (event);
+      ASSERT (!ctx->has_reply);
+      ctx->has_reply = true;
+      bson_copy_to (mongoc_apm_command_succeeded_get_reply (event),
+                    &ctx->reply);
+   }
+}
+
+static void
+prose_test_2 (void *ctx)
+{
+   mongoc_client_t *client;
+   mongoc_database_t *db;
+   mongoc_collection_t *coll;
+   mongoc_apm_callbacks_t *callbacks;
+   prose_test_2_apm_ctx_t apm_ctx = {0};
+   bool ret;
+   bson_t reply, reply_errInfo, observed_errInfo;
+   bson_error_t error = {0};
+
+   client = test_framework_new_default_client ();
+   db = get_test_database (client);
+   coll = get_test_collection (client, "coll");
+
+   /* don't care if ns not found. */
+   (void) mongoc_collection_drop (coll, NULL);
+
+   ret = mongoc_database_create_collection (
```

Comment:
This returns a `mongoc_collection_t*` that must be destroyed.

Reply:
Oh, that might explain one of the odder test failures I came across earlier. I'll update this.